include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,123 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+NBL_CONSTEXPR uint32_t BucketsPerThread = ceil((float) BucketCount / WorkgroupSize);
+
+groupshared uint32_t prefixScratch[BucketCount];
+
+struct ScratchProxy
+{
+    uint32_t get(const uint32_t ix)
+    {
+        return prefixScratch[ix];
+    }
+    void set(const uint32_t ix, const uint32_t value)
+    {
+        prefixScratch[ix] = value;
+    }
+
+    void workgroupExecutionAndMemoryBarrier()
+    {
+        nbl::hlsl::glsl::barrier();
+    }
+};
+
+static ScratchProxy arithmeticAccessor;
+
+groupshared uint32_t sdata[BucketCount];
```

Comment:
1. we have a hard policy of NO GLOBALS, so all this goes into userspace (example code)
2. perf-tip, you can alias `prefixScratch` to `sdata` (use the same scratch accessor for histogramming and prefix summing) as long as you "load" the histogram values before you overwrite the bottom of the scratch and store the results of the scan that would write to the bottom of the scratch AFTER the final prefix sum

So your `atirhmeticAccessor` and `prefixScratch` are not needed

Reply:
Done